internal/clients/aws.go

```diff
@@ -6,62 +6,184 @@ package clients
-	return func(ctx context.Context, client client.Client, mg resource.Managed) (terraform.Setup, error) {
-		cfg, err := GetAWSConfig(ctx, client, mg)
+func SelectTerraformSetup(version, providerSource, providerVersion string) terraform.SetupFn {
+	return func(ctx context.Context, c client.Client, mg resource.Managed) (terraform.Setup, error) {
+		pc := &v1beta1.ProviderConfig{}
+		if err := c.Get(ctx, types.NamespacedName{Name: mg.GetProviderConfigReference().Name}, pc); err != nil {
+			return terraform.Setup{}, errors.Wrapf(err, "cannot get referenced Provider: %s", mg.GetProviderConfigReference().Name)
+		}
+		ps := terraform.Setup{
+			Version: version,
+			Requirement: terraform.ProviderRequirement{
+				Source:  providerSource,
+				Version: providerVersion,
+			},
+		}
+		account, err := getAccountId(ctx, c, mg)
 		if err != nil {
-			return terraform.Setup{}, errors.Wrap(err, "cannot get AWS config")
+			return terraform.Setup{}, errors.Wrap(err, "cannot get account id")
 		}
-		if cfg.Region == "" && mg.GetObjectKind().GroupVersionKind().Group == "iam.aws.upbound.io" {
-			cfg.Region = "us-east-1"
+		ps.ClientMetadata = map[string]string{
+			keyAccountId: account,
 		}
-		creds, err := cfg.Credentials.Retrieve(ctx)
+
+		if len(pc.Spec.AssumeRoleChain) > 1 || pc.Spec.Endpoint != nil {
+			err = DefaultTerraformSetupBuilder(ctx, c, mg, &ps)
+			if err != nil {
+				return terraform.Setup{}, errors.Wrap(err, "cannot build terraform configuration")
+			}
+		} else {
+			err = pushDownTerraformSetupBuilder(ctx, c, mg, pc, &ps)
+			if err != nil {
+				return terraform.Setup{}, errors.Wrap(err, "cannot build terraform configuration")
+			}
+		}
+		return ps, nil
+	}
+}
+
+func pushDownTerraformSetupBuilder(ctx context.Context, c client.Client, mg resource.Managed, pc *v1beta1.ProviderConfig, ps *terraform.Setup) error { //nolint:gocyclo
+	if len(pc.Spec.AssumeRoleChain) > 1 || pc.Spec.Endpoint != nil {
+		return errors.New("shared scheduler cannot be used because the length of assume role chain array " +
+			"is more than 1 or endpoint configuration is not nil")
+	}
+
+	cfg, err := getAWSConfig(ctx, c, mg)
+	if err != nil {
+		return errors.Wrap(err, "cannot get AWS config")
+	}
+	ps.Configuration = map[string]any{
+		keyRegion: cfg.Region,
+	}
+
+	switch s := pc.Spec.Credentials.Source; s { //nolint:exhaustive
+	case authKeyWebIdentity:
+		if pc.Spec.Credentials.WebIdentity == nil {
+			return errors.New(`spec.credentials.webIdentity of ProviderConfig cannot be nil when the credential source is "WebIdentity"`)
+		}
+		if pc.Spec.Credentials.WebIdentity.RoleARN == nil {
+			return errors.New(`spec.credentials.webIdentity.roleARN of ProviderConfig cannot be nil when the credential source is "WebIdentity"`)
+		}
+		ps.Configuration[keyAssumeRoleWithWebIdentity] = map[string]any{
+			keyRoleArn:              pc.Spec.Credentials.WebIdentity.RoleARN,
+			keySessionName:          pc.Spec.Credentials.WebIdentity.RoleSessionName,
```

Comment:
If the `pc.Spec.Credentials.WebIdentity.RoleSessionName` is an empty string, Terraform complains about it:
```shell
"Error: expected length of assume_role_with_web_identity.0.session_name to be in the range (2 - 64), got "
```

One solution could be converting the type of `pc.Spec.Credentials.WebIdentity.RoleSessionName` to `*string` but we need to make sure upgrades work smoothly (expecting empty strings to be converted to `nil` because of the `omitempty` JSON serialization tag). 

Another option could be not to put this key into the map if the value is an empty string.

Reply:
For the first iteration of these improvements, I preferred adding an empty string check. I think in the next iteration we may consider converting the field to `*string`.

Reply:
Thanks @sergenyalcin. Successfully validated the following two additional provider configurations with the fix to this issue using the package `ulucinar/provider-aws-amd64:v0.31.0-ca2f21ca`:
- WebIdentity
- WebIdentity with an AssumeRole chain of length 1
- WebIdentity with an AssumeRole chain of length 2